server.go

```diff
@@ -97,16 +93,44 @@ func (s *Server) ServeHTTPS(addr, keyfile, certfile string, ready *chan bool) er
 	if s.listener, err = listenTLS(addr, keyfile, certfile); err != nil {
 		return err
 	}
+	s.tls = true
+	fmt.Printf("Listen http on %s\n", addr)
+	return s.doServe(ready)
+}
+
+func (s *Server) doServe(ready *chan bool) error {
+	// A dirty trick to associate a connection with the http.Request it
+	// contains. In "net/http/server.go", handler will be called
+	// immediately after ConnState changed to StateActive, so it's safe to
+	// loop through all elements in a channel to find a match remote addr.
+	q := make(chan net.Conn, 10)
 
-	// Set up server
 	proxy := http.HandlerFunc(
 		func(w http.ResponseWriter, req *http.Request) {
+			for c := range q {
+				if c.RemoteAddr().String() == req.RemoteAddr {
```

Comment:
I think this code will loop forever if there is some mismatch.  context.Get
has no chance to return nil.

On Thursday, October 22, 2015, Ulysses Aalto <notifications@github.com>
wrote:

> In server.go
> <https://github.com/getlantern/chained-server/pull/10#discussion_r42761778>
> :
>
> >  	proxy := http.HandlerFunc(
> >  		func(w http.ResponseWriter, req *http.Request) {
> > +			for c := range q {
> > +				if c.RemoteAddr().String() == req.RemoteAddr {
>
> Exactly as you said it, it will be hard to spot the cause if it there is a
> problem. Since this is the best we got before redesigning the whole go http
> server req/resp architecture, we should keep in mind the kind of errors
> this would cause. I think a test of nullity for context.Get in uidfilter.go
> would help, what do you think?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/getlantern/chained-server/pull/10/files#r42761778>.
>